upstream: outlier detector - distinguish upstream from internal errors

cpakulski opened the pull request:
Description: Outlier detector has specific handlers for 5xx errors. Internal errors, like connect failures, timeout, resets have been mapped to 5xx errors and handled by the same logic in outlier detector as upstream's 5xx errors. This PR introduces separate handlers for internal errors (connect failures, timeout, resets). New internal counters and new config items have been added. 

Risk Level: Low

Testing: Unit tests.

Docs Changes: Yes - new config items have been added to config reference and section about outlier detector.

Release Notes: N/A
Fixes #3643

cpakulski commented:
I believe that I have implemented that feature correctly. In most places it was just copy and paste of outlier's methods for 5xx errors. Only in https://github.com/envoyproxy/envoy/blob/09bba398f844fe2eab8fc57177c6bdac98e6d7b0/source/common/router/router.cc#L478 and https://github.com/envoyproxy/envoy/blob/09bba398f844fe2eab8fc57177c6bdac98e6d7b0/source/common/router/router.cc#L1000 I had to add calls to report failure and success.
All test cases are passing OK. However, there are few things I am not sure about:
-	I did not modify v1 config. It means that logic to differentiate connect errors and 5xx error is there, but works only with defaults and cannot be modified via v1 config.
-	I think that documentation is not clear enough to explain differences between gateway errors, 5xx errors and connect errors. I can extend docs, but I am not sure what cases are handled by gateway errors.
-	In router_test.cc I had to replace calls to outlier detector’s handler for 5xx with calls to handler for connect errors. However, in many places there are references to http headers, like "x-envoy-retry-on", "5xx" and I am not sure if they still make sense when connect errors are processes by a handler for connect failures, not 5xx handlers. Like it is done here: https://github.com/envoyproxy/envoy/blob/09bba398f844fe2eab8fc57177c6bdac98e6d7b0/test/common/router/router_test.cc#L848-L862. I would need some guidance here what is the purpose of "x-envoy-retry-on" and whether it’s use should be extended for “connect” type errors.

mattklein123 commented on include/envoy/upstream/outlier_detection.h:
```diff
@@ -82,6 +82,9 @@ class DetectorHostMonitor {
    */
   virtual double successRate() const PURE;
+
+  virtual void connectFailure() PURE;
```
doc comments please

mattklein123 commented on source/common/router/router.cc:
```diff
@@ -475,9 +475,7 @@ void Filter::onUpstreamReset(UpstreamResetType type,
-      upstream_host->outlierDetector().putHttpResponseCode(
-          enumToInt(type == UpstreamResetType::Reset ? Http::Code::ServiceUnavailable
-                                                     : timeout_response_code_));
+      upstream_host->outlierDetector().connectFailure();
```
tcp_proxy also supports outlier detection. I think that should be modified with these new methods?

mattklein123 commented on api/envoy/api/v2/cluster/outlier_detection.proto:
```diff
@@ -76,4 +76,14 @@ message OutlierDetection {
+
+  // The number of consecutive connect failures before ejection
+  // occurs. Defaults to 5.
+  google.protobuf.UInt32Value consecutive_connect_failure = 12;
```
Unfortunately I think this is more complicated than "connect failure" since some of the reasons for "connection failures" can be internal circuit breaking. We should make a more generic name here and clearly document how this is basically all errors that are internally generated by Envoy (failure to TCP connection, TLS handshake, overflow, timeout, etc.).

One special case that needs considering is when upstream actually *does* send an HTTP reset. How should this be handled? IMO it should count as an upstream real service unavailable error. Anyway needs thinking about.

Once we lock down the nomenclature here I think we will need to update the entire PR to use the new wording.

cpakulski replied:
> Unfortunately I think this is more complicated than "connect failure" since some of the reasons for "connection failures" can be internal circuit breaking. We should make a more generic name here and clearly document how this is basically all errors that are internally generated by Envoy (failure to TCP connection, TLS handshake, overflow, timeout, etc.).

@mattklein123 I can definitely address those cases. Outlier detector works on the principle that errors are reported to an outlier implementation attached to a host. I guess that other errors you mentioned: TCP connection, TLS handshake, overflow, timeout are currently not reported to host's outlier detector, correct? 

> We should make a more generic name...

What is your suggestion for naming those types of errors?

mattklein123 replied:
> I guess that other errors you mentioned: TCP connection, TLS handshake, overflow, timeout are currently not reported to host's outlier detector, correct?

Yes they are reported. These generally come through as "resets" of various types.

> What is your suggestion for naming those types of errors?

I'm not sure of the best name, but something along the lines of "local origin error" vs. "remote origin error" or something like that is how I would think about it.

cpakulski replied:
Thanks. I will look at TCP connection and TLS handshake first and will come up with a proposal.

mattklein123 replied:
If we are looking for names here, I would do something like: "local_origin" vs. "connect" and describe what that means. I think "local_origin" is a wrapper for circuit breaking, connect failure, etc.

mattklein123 commented:
Marking is waiting. 

/wait

cpakulski commented:
Summary of this PR push:
- removed mapping of connection errors, server errors to HTTP codes. This change will affect reporting of connection problems by router, tcp_proxy and redis_proxy.
- updated documentation to better explain how outlier detector uses its failure counters. I hope it will also fix  #3730 

It is still not in the final form. Just want to get a feedback whether direction is good. The following changes are still required:
- I still used word _connect_ to describe connection problems. As noted by @mattklein123 (above in this PR) better name should be used, but I could not come up with a better naming.
- Outlier detector used to have only two counters: consecutive_5xx and consecutive_gateway_failures. Now there are 2 more: consecutive_connect_failure and consecutive_server_request_failure. Handling of those counters is almost identical and there is lots of code repetition. I think we should collapse the code to use a common class or template.
- SuccessRate algorithm rides on top of HTTP error code. Because tcp_proxy and redis_proxy errors were mapped to HTTP code, SuccessRate logic automatically applied to those filters as well. Now, tcp_proxy, redis_proxy and router's connect failures have their own handlers for connect errrors and SuccessRate logic does not apply there. I can add it if it makes sense.

Just want to get a feedback, if I should continue with those changes or drop them.

htuch commented:
@cpakulski will you be making the success rate parameters (internal vs. connect/request derived failures) a tunable option?

mattklein123 commented on api/envoy/api/v2/cluster/outlier_detection.proto:
```diff
@@ -76,4 +76,24 @@ message OutlierDetection {
+  google.protobuf.UInt32Value enforcing_consecutive_connect_failure = 13
+      [(validate.rules).uint32.lte = 100];
+
+  // The number of consecutive server transaction failures before ejection occurs.
```
Just from reading this documentation, I'm not sure what a server transaction failure is. I will keep reading the PR but can we update that to make it more clear?

Edit: OK So the idea here is this is an upstream server failure for non-HTTP servers? Is that right? If so I think that makes sense.

cpakulski replied:
That is correct. Upstream server failure would be a situation when connection was OK, but for example redis DB transaction failed. Server could not understand what it received or parsing of reply failed.

mattklein123 commented on docs/root/intro/arch_overview/outlier.rst:
```diff
@@ -10,6 +10,17 @@ such as consecutive failures, temporal success rate, temporal latency, etc. Outl
 <arch_overview_health_checking>`. *Passive* and *active* health checking can be enabled together or
 independently, and form the basis for an overall upstream health checking solution.
+Outlier detection is part of :ref:`cluster configuration <envoy_api_msg_cluster.OutlierDetection>`, 
+but it needs network filters to report errors, timeouts, resets. Currently the following filters support
```
"network or HTTP filters" ?

cpakulski replied:
AFAIK only HTTP router, tcp_proxy and redis_proxy provide feedback to outlier detector. The latter two do not really talk http, so I called them "network".

mattklein123 commented on docs/root/intro/arch_overview/outlier.rst:
```diff
@@ -113,7 +146,7 @@ action
 type
   If ``action`` is ``eject``, specifies the type of ejection that took place. Currently type can
-  be one of ``5xx``, ``GatewayFailure`` or ``SuccessRate``.
+  be one of ``5xx``, ``GatewayFailure``, ``SuccessRate``, ``ConnectFailure`` or ``MalformedPayload``.
```
What is "MalformedPayload" ?

cpakulski replied:
My bad. Thanks for catching it. This is how I originally called "upstream server failure", but later on changed naming.

mattklein123 commented:
> SuccessRate algorithm rides on top of HTTP error code. Because tcp_proxy and redis_proxy errors were mapped to HTTP code, SuccessRate logic automatically applied to those filters as well. Now, tcp_proxy, redis_proxy and router's connect failures have their own handlers for connect errrors and SuccessRate logic does not apply there. I can add it if it makes sense.

Yes I think we definitely need it, though for my understanding, what was the issue with continuing to share the HTTP logic? Clarity or something else?

cpakulski commented:
>  will you be making the success rate parameters (internal vs. connect/request derived failures) a tunable option?

@htuch I will try to do it such way that there is one config parameter set for success rate and those settings will apply to all kind of failures. I am still not sure whether to treat failure counters separately or in aggregate. So, for example it a host fails because of http 5xx errors and also occasionally fails because of internal errors, should Success Rate treat those errors separately or sum them together. Based on my understanding of #3643, they should be treated separately.

cpakulski commented:
> > SuccessRate algorithm rides on top of HTTP error code. Because tcp_proxy and redis_proxy errors were mapped to HTTP code, SuccessRate logic automatically applied to those filters as well. Now, tcp_proxy, redis_proxy and router's connect failures have their own handlers for connect errrors and SuccessRate logic does not apply there. I can add it if it makes sense.
> 
> Yes I think we definitely need it, though for my understanding, what was the issue with continuing to share the HTTP logic? Clarity or something else?

There were few reasons. First was that in order to distinguish internal errors from upstream server replies, new way of reporting such events to outlier detector was required. Those events had separate counters for failures. Originally I added _connectSuccess/Failure_ calls to router, which updated newly added counters. Then I added the new API calls to tcp_proxy and redis proxy. At that moment mapping to http code stopped being used.
The second reason was more of the clarity. By reading documentation I completely did not understand how the outlier detector's mechanism worked. Only by reading the code, the logic became clear. But from a user point of view, it is difficult to make a connection between  tcp_connect failures or redis connect failures to 5xx http error codes, and configuration uses http error codes.

mattklein123 commented:
> Based on my understanding of #3643, they should be treated separately.

Agreed separately.

> The second reason was more of the clarity.

OK that's fine. The use of HTTP codes previous was kind of a hack. With that said, we need to keep the functionality we had before. Meaning, we still must support SR outlier detection for redis and TCP.

> I know I still used connect name to describe it, will change it to something else (internal?).

Personally I would use that I already suggested which is "local_origin." If you trace through the code, there are many things that can cause this: timeouts, circuit breakers, connect errors, etc. I think as long as it's well documented, this will be clear to people.

Regarding "server failure" vs. "HTTP success rate" I'm still not sure if it's worth configuring different options for this, since logically it's really the same thing. Can we somehow merge them so we only have two concepts: "local origin errors" and "upstream errors?"

cpakulski commented:
@mattklein123 Thanks for your patience and help in explaining this. I think we are on the same page now. 

> Regarding "server failure" vs. "HTTP success rate" I'm still not sure if it's worth configuring different options for this, since logically it's really the same thing. Can we somehow merge them so we only have two concepts: "local origin errors" and "upstream errors?"

I will try to code it that way.

htuch commented:
+1 on local origin as name to distinguish from actual backends. 

> @htuch I will try to do it such way that there is one config parameter set for success rate and those settings will apply to all kind of failures. I am still not sure whether to treat failure counters separately or in aggregate. So, for example it a host fails because of http 5xx errors and also occasionally fails because of internal errors, should Success Rate treat those errors separately or sum them together. Based on my understanding of #3643, they should be treated separately.

Yep, separately.

cpakulski commented:
Work in progress.

cpakulski commented:
Update: testing success rate for internal origin errors.

cpakulski commented:
Update: finishing testing and updating docs.

dnoe commented:
Let me know whenever you are ready for another review pass. Thanks!

mattklein123 commented:
@dnoe I think this is ready for another pass. LMK once you are done.

cpakulski commented:
Summary of the latest push:
- local origin errors are reported from tcp_proxy, router and redis_proxy modules. There is a separate counter for them and 2 extra config items to tell how many consecutive local origin errors make a host candidate for ejection and what is the probability of ejection. This is standard stuff and copy-and-paste of 5xx errors logic.
- For success rate monitor things are a bit more complicated, because success rate requires more data like buckets, accumulators, etc. I gathered all that data in new class SuccessRateMonitor and created two instances for external origin and for local origin. The rest is as before - calculations and ejections are timer based.
- configuration: success rate is configured with 4 params: minimum_hosts, request_volume, stdev_factor and enforcing.... The first 3 config parameters are shared for both external and local origin success rate monitors. Specific success rate (external or local) can be switched off by specifying 0 for _enforcing_success_rate_ or _enforcing_local_origin_success_rate_ parameter respectively. The only scenario which cannot be configured is to enable success rate for both external and local errors and specify different values for minimum_hosts, request_volume and stdev_factor.

mattklein123 commented:
@cpakulski looks like needs a master merge. @dnoe post vacation friendly ping.

/wait

cpakulski commented:
@dnoe Rebased with master and ready for review.

mattklein123 commented:
@dnoe friendly ping: will you have time to review this?

dnoe commented on api/envoy/api/v2/cluster/outlier_detection.proto:
```diff
@@ -78,4 +78,21 @@ message OutlierDetection {
+  // The % chance that a host will be actually ejected when an outlier status
+  // is detected through consecutive locally originated failures. This setting can be
+  // used to disable ejection or to ramp it up slowly. Defaults to 100.
+  google.protobuf.UInt32Value enforcing_consecutive_local_origin_failure = 13
```
What do you think about making "percent" part of this name so it's clear it is a percent? Likewise for below.

cpakulski replied:
That is probably good idea, but I just followed a convention. All names starting with enforcing are represented as percentages: https://github.com/envoyproxy/envoy/blob/master/api/envoy/api/v2/cluster/outlier_detection.proto#L38-L41 and https://github.com/envoyproxy/envoy/blob/master/api/envoy/api/v2/cluster/outlier_detection.proto#L43-L46

dnoe replied:
Aha, SGTM. Following the convention is good.

dnoe commented on api/envoy/api/v2/cluster/outlier_detection.proto:
```diff
@@ -78,4 +78,21 @@ message OutlierDetection {
+  // The % chance that a host will be actually ejected when an outlier status
+  // is detected through success rate statistics for locally originated errors.
+  // This setting can be used to
+  // disable ejection or to ramp it up slowly. Defaults to 100.
```
nit: fix up the premature new line in the comment and let fix-format do its thing to wrap the comment.

cpakulski replied:
Done.

dnoe commented on source/common/upstream/outlier_detection_impl.cc:
```diff
@@ -330,24 +408,28 @@ void DetectorImpl::onConsecutiveErrorWorker(HostSharedPtr host, EjectionType typ
 }
 
-Utility::EjectionPair Utility::successRateEjectionThreshold(
+std::tuple<double, double> DetectorImpl::successRateEjectionThreshold(
```
Any particular reason why you are using `std::tuple` rather than `std::pair` for this two element tuple? In general I think both approaches suffer a bit from being opaque about what the pair of numbers is, maybe a typedef or simple struct is appropriate? What do you think?

cpakulski replied:
Converted to use simple struct.

dnoe commented on source/common/upstream/outlier_detection_impl.cc:
```diff
@@ -418,15 +501,17 @@ void DetectorImpl::processSuccessRateEjections() {
         success_rate_sum, valid_success_rate_hosts, success_rate_stdev_factor);
-    success_rate_average_ = ejection_pair.success_rate_average_;
-    success_rate_ejection_threshold_ = ejection_pair.ejection_threshold_;
+    double success_rate_ejection_threshold = std::get<1>(success_rate_nums_[monitor_type]);
```
nit: `const double`

cpakulski replied:
Done.

dnoe commented on source/common/upstream/outlier_detection_impl.cc:
```diff
@@ -418,15 +501,17 @@ void DetectorImpl::processSuccessRateEjections() {
         stats_.ejections_success_rate_.inc(); // Deprecated.
-        stats_.ejections_detected_success_rate_.inc();
-        ejectHost(host_success_rate_pair.host_, EjectionType::SuccessRate);
+        EjectionType type = host_monitors_[host_success_rate_pair.host_]
```
`const EjectionType`

cpakulski replied:
Done.

dnoe commented on source/common/upstream/outlier_detection_impl.h:
```diff
@@ -130,19 +159,45 @@ class DetectorHostMonitorImpl : public DetectorHostMonitor {
+  // counters for local origin failures
+  std::atomic<uint32_t> consecutive_local_origin_failure_{0};
+
+  std::map<SuccessRateMonitorType, std::unique_ptr<SuccessRateMonitor>> success_rate_monitors_;
```
Did you consider `absl::flat_hash_map` for this?

cpakulski replied:
Thanks for suggestion. Converted std::map to absl::flat_map_hash.

dnoe commented on source/common/upstream/outlier_detection_impl.h:
```diff
@@ -189,6 +248,11 @@ class DetectorConfig {
   uint64_t enforcingConsecutive5xx() { return enforcing_consecutive_5xx_; }
   uint64_t enforcingConsecutiveGatewayFailure() { return enforcing_consecutive_gateway_failure_; }
   uint64_t enforcingSuccessRate() { return enforcing_success_rate_; }
+  uint64_t consecutiveLocalOriginFailure() { return consecutive_local_origin_failure_; }
```
I might be missing something obvious since it isn't on any of the other existing methods here, but this is a `const` method, right?

dnoe commented on source/common/upstream/outlier_detection_impl.h:
```diff
@@ -219,13 +286,34 @@ class DetectorImpl : public Detector, public std::enable_shared_from_this<Detect
+   * @param success_rate_sum is the sum of the data in the success_rate_data vector.
+   * @param valid_success_rate_hosts is the vector containing the individual success rate data
+   *        points.
+   * @return std::tuple<double, double>.
```
Unless you switch to using a struct with named fields, please make sure we're clearly documenting what the two doubles in the tuple actually are.

cpakulski replied:
Using simple struct now with meaningful data members' names and comments.

dnoe commented on source/common/upstream/outlier_detection_impl.h:
```diff
@@ -255,8 +344,9 @@ class DetectorImpl : public Detector, public std::enable_shared_from_this<Detect
-  double success_rate_average_;
-  double success_rate_ejection_threshold_;
+
+  std::map<DetectorHostMonitor::SuccessRateMonitorType, std::tuple<double, double>>
```
Maybe `absl::flat_hash_map` here too.

cpakulski replied:
Thanks for suggestion. Converted std::map to absl::flat_map_hash.

dnoe commented on source/server/http/admin.cc:
```diff
@@ -360,12 +379,19 @@ void AdminImpl::writeClustersAsJson(Buffer::Instance& response) {
 #undef SET_HEALTH_FLAG
 
-        double success_rate = host->outlierDetector().successRate();
+        double success_rate = host->outlierDetector().successRate(
```
nit: `const double`

cpakulski replied:
Done.

cpakulski replied:
Actually, success_rate is used later on in write operation. Must stay as is.

cpakulski commented:
@dnoe I have addressed all your comments. Rebasing now, as there are few conflicting files.

cpakulski commented:
@dnoe I am done with rebasing from master. #5517 complicated few things, as it removed ejection types defined as enums and introduced ejection types defined in protobufs https://github.com/envoyproxy/envoy/blob/master/api/envoy/data/cluster/v2alpha/outlier_detection_event.proto#L49-L56. I just had to refactor my code to use new types from protobufs. Logic did not change since your last code review, just new types are used.

cpakulski commented:
@dnoe Could you please do "second" first pass review?

dnoe commented:
Whoops, sorry I missed the notification on this one. I will take a look tonight, thanks for your patience.

dnoe commented:
@mattklein123 for final approval

mattklein123 commented:
@cpakulski will review today. Can you merge master since it's been a while? Thank you.

cpakulski commented:
@mattklein123 rebased.

mattklein123 commented on api/envoy/admin/v2alpha/clusters.proto:
```diff
@@ -43,6 +43,19 @@ message ClusterStatus {
+
+  // The success rate threshold used in the last interval for locally originated failures.
+  // The threshold is used to eject hosts based on their success rate. See
+  // :ref:`Cluster outlier detection <arch_overview_outlier_detection>` statistics
```
nit: end sentence with period. Can you fix elsewhere in this section if you see similar instances)

mattklein123 replied:
Per above, maybe reference `split_external_local_origin_errors` and say that this field won't do anything if that is not true?

mattklein123 commented on source/common/config/cds_json.cc:
```diff
@@ -97,6 +97,12 @@ void CdsJson::translateOutlierDetection(
   JSON_UTIL_SET_INTEGER(json_outlier_detection, outlier_detection, success_rate_minimum_hosts);
   JSON_UTIL_SET_INTEGER(json_outlier_detection, outlier_detection, success_rate_request_volume);
   JSON_UTIL_SET_INTEGER(json_outlier_detection, outlier_detection, success_rate_stdev_factor);
+  JSON_UTIL_SET_INTEGER(json_outlier_detection, outlier_detection,
```
Please revert this file, we should no longer be supporting any v1 config changes.

mattklein123 commented on api/envoy/api/v2/cluster/outlier_detection.proto:
```diff
@@ -80,4 +80,20 @@ message OutlierDetection {
+  // The % chance that a host will be actually ejected when an outlier status
+  // is detected through success rate statistics for locally originated errors.
+  // This setting can be used to disable ejection or to ramp it up slowly. Defaults to 100.
+  google.protobuf.UInt32Value enforcing_local_origin_success_rate = 14
```
I'm a little confused here. There is now an enforcing setting for local origin success rate, but no setting for local origin success rate? Where is the local origin success rate configured?

At a high level, I'm confused as to whether this PR is going to have user visible effect for people that make no config changes. Will it? Either way, can you add a release note that describes everything at a high level?

cpakulski replied:
Thanks for reviewing my changes. I will address your other comments, but we have to agree on configuration strategy first. Success rate algorithm uses 3 parameters: success_rate_minimum_hosts, success_rate_request_volume and success_rate_stdev_factor. The first 2 are self-explanatory. The stdev is used to calculate the success threshold. If a node's success rate is below threshold it will be candidate for ejection. The final decision about ejecting is determined by enforcing_success_rate parameter: 0 means - never eject and 100 means always eject. Value in between will be used by random number generator to provide "probability" of ejection. 
In the code there are now 2 instances of SuccessRate classes. One tracks failure rate of external events and the other tracks failure rate of local origin events. To avoid config pollution, the mentioned 3 parameters: success_rate_minimum_hosts, success_rate_request_volume and success_rate_stdev_factor are used to initialize both instances of SuccessRate classes. There is however enforcing parameter for each type of SuccessRate (external and local origin). Sure, I could repeat those 3 parameters and changed their name to local_origin_xxx, but will anyone ever use it?
With the current parameters one can configure:
- turn off one type of Success Rate (set enforcing of one type to zero)
- turn on both (set enforcing of both types to values greater than zero)
- make probability of ejection for one type of SuccessRate to be more likely to happen

The only scenario which cannot be configured is to have a different minimum number hosts, volume or ejection threshold for each SuccessRate type. I doubt anyone would try to configure such scenario, but I can duplicate those config items with local_origin_ prefix if you think it makes sense.

> I'm confused as to whether this PR is going to have user visible effect for people that make no config changes.

I think that when a user uses just default values there will be no visible effect. I wrote "I think" because previously failures went to one bucket and now go into two buckets (external and local origin) and if volume of requests is small, one can imagine corner case that the number of failures in each bucket will not be sufficient to eject a host. I have to experiment with this more. 

If a  user modified enforcing parameter, that setting will now apply only for external origin SuccessRate. For internal origin SuccessRate default of 100 will be used.

mattklein123 replied:
OK thanks for the explanation. The first part about how you only have an enforcing setting makes sense to me. My strong preference here would be to figure out a way of making it so that there is no impact to existing users if they don't change any config.

I'm wondering if we should have this new behavior be opt-in? For example, there could be a parameter along the lines of `split_external_local_origin_errors` or something like that. If this is not true (default) there is no change in behavior, otherwise we split and then have different settings? I actually think most users are not going to want this split at all.

cpakulski replied:
OK. I can add this param to provide no-change-in-behavior for existing users. As for the name it could be `split_external_local_origin_error`or maybe `distinguish_local_origin_errors`. I will use your suggestion. Default would be `false`. 
I will try implement it such way that depending on the value of that parameter call chain will be set up differently, so that flag does not have to be checked each time a packet is processed.
Not sure when I will finish it, as I have to travel within new few days, but this is my main focus now.

mattklein123 replied:
Thanks, I think this will provide the best configuration experience for most users.

cpakulski replied:
Agree. Sorry that I cannot finish it very soon, but as I said before I will be travelling.

mattklein123 commented on api/envoy/data/cluster/v2alpha/outlier_detection_event.proto:
```diff
@@ -52,7 +52,12 @@ enum OutlierEjectionType {
   // Runs over aggregated success rate statistics from every host in cluster
-  SUCCESS_RATE = 2;
+  SUCCESS_RATE_EXTERNAL_ORIGIN = 2;
+  // Consecutive local origin failures: Connection failures, resets, etc
```
In this docs, can you link "local origin" to the arch overview docs and a section that clearly defines them? I'm concerned that users are going to find all of these config knobs very confusing (I admit I do).

cpakulski commented:
@mattklein123 I am ready with the code. Can you reopen this PR, please? All relevant comments about this issue are here. Thx.

mattklein123 commented:
@cpakulski reopened.

/wait

mattklein123 commented on api/envoy/admin/v2alpha/clusters.proto:
```diff
@@ -28,8 +28,8 @@ message ClusterStatus {
 
-  // The success rate threshold used in the last interval. The threshold is used to eject hosts
-  // based on their success rate. See
+  // The success rate threshold used in the last interval for externally generated failures.
```
Please add a release note for all the changes in this PR.

mattklein123 replied:
Also, should this doc reference the `split_external_local_origin_errors` setting so it's clear what this field sets depending on the setting? Feel free to link back to the arch docs if that would be easier.

cpakulski replied:
Release note added.

mattklein123 commented on api/envoy/api/v2/cluster/outlier_detection.proto:
```diff
@@ -80,4 +80,30 @@ message OutlierDetection {
+  // consecutive_local_origin_failure, consecutive_local_origin_failure
+  // enforcing_consecutive_local_origin_failure and enforcing_local_origin_success_rate.
+  // Defaults to false.
+  google.protobuf.BoolValue split_external_local_origin_errors = 12;
```
This can be a standard `bool` value.

cpakulski replied:
I tried to change it to _bool_, but compiler complains that
PROTOBUF_GET_WRAPPED_OR_DEFAULT macro cannot resolve _has_##field_name()_ which maps to non-existing _has_split_external_local_origin_errors_.
I can investigate it further if it is important not to use _google.protobuf.BoolValue_.

mattklein123 replied:
You shouldn't need any default for a `bool` as the default is false. That's why we should use it here.

cpakulski replied:
OK. got it. Thanks!

mattklein123 commented on api/envoy/api/v2/cluster/outlier_detection.proto:
```diff
@@ -80,4 +80,30 @@ message OutlierDetection {
+
+  // Determines whether to distinguish local origin failures from external errors. If set to true
+  // the following configuration parameters are taken into account:
+  // consecutive_local_origin_failure, consecutive_local_origin_failure
```
Please ref link all fields. Same elsewhere in this change.

cpakulski replied:
Done. Here and everywhere else. Thanks for catching it.

mattklein123 commented on docs/root/configuration/cluster_manager/cluster_stats.rst:
```diff
@@ -130,10 +130,14 @@ statistics will be rooted at *cluster.<name>.outlier_detection.* and contain the
   ejections_detected_consecutive_5xx, Counter, Number of detected consecutive 5xx ejections (even if unenforced)
-  ejections_enforced_success_rate, Counter, Number of enforced success rate outlier ejections
-  ejections_detected_success_rate, Counter, Number of detected success rate outlier ejections (even if unenforced)
+  ejections_enforced_success_rate, Counter, Number of enforced success rate outlier ejections for external failures
```
This is not always true if we are not splitting. Same elsewhere. Can you audit docs to make this more clear across all the docs in case there are other similar places?

cpakulski replied:
Corrected.

cpakulski commented:
I do not know why _release_ tests are failing. It only says g++ internal error. Is it a known problem?

mattklein123 commented on docs/root/intro/arch_overview/outlier.rst:
```diff
@@ -10,6 +10,39 @@ such as consecutive failures, temporal success rate, temporal latency, etc. Outl
+configuration items. For example, if connection to an upstream HTTP server fails twice because of timeout and 
+then, after successful connection, the server returns error code 500, the total error count will be 3. 
+
+Outlier detection may also be configured to distinguish locally originated (connection) errors from externally originated (transaction) errors. 
```
All locally originated errors are not connection errors, right? Meaning, if a request times out it will fail in a locally originated sense, but it's not a connection level error? Is that right? If so can you clarify here and elsewhere?

cpakulski replied:
Did my best to explain locally originated errors.

mattklein123 commented on docs/root/intro/version_history.rst:
```diff
@@ -22,6 +22,7 @@ Version history
   in full by the router. This ensures that the per try timeout does not account for slow
   downstreams and that will not start before the global timeout.
 * upstream: added :ref:`upstream_cx_pool_overflow <config_cluster_manager_cluster_stats>` for the connection pool circuit breaker.
+* outlier_detector: added configuration :ref:`outlier_detection.split_external_local_origin_errors<envoy_api_field_cluster.OutlierDetection.split_external_local_origin_errors>` to distinguish locally and externally generated errors. See :ref:`arch_overview_outlier_detection` for full details.
```
nit: alpha order

cpakulski replied:
Done.

cpakulski commented on api/envoy/admin/v2alpha/clusters.proto:
```diff
@@ -28,9 +28,9 @@ message ClusterStatus {
-  // The success rate threshold used in the last interval. The threshold is used to eject hosts
-  // based on their success rate. See
-  // :ref:`Cluster outlier detection <arch_overview_outlier_detection>` statistics
+  // The success rate threshold used in the last interval for externally generated failures.
```
Corrected. Thanks for catching it.

mattklein123 commented on source/common/router/router.cc:
```diff
@@ -1197,6 +1202,8 @@ void Filter::UpstreamRequest::onPoolReady(Http::StreamEncoder& request_encoder,
                                           Upstream::HostDescriptionConstSharedPtr host) {
   ENVOY_STREAM_LOG(debug, "pool ready", *parent_.callbacks_);
 
+  host->outlierDetector().putResult(Upstream::Outlier::Result::CONNECT_SUCCESS);
```
This doesn't necessarily mean that a new connection was made since it may be using a preexisting connection. I'm not sure we want to account for this here? I guess you need this if you are going to tracking connection failures like you do above? Thoughts?

cpakulski replied:
I think I need it here, even when it may be using preexisting connection. Let us consider situation when there is 50 local origin errors (ex. port on upstream server was not opened). After that things are OK and subsequent 1000 requests are fine. Even when those 1000 requests use one preexisting connection, the ration of success to failures should be 1000 to 50, so we should count those successes here.

mattklein123 commented on source/common/upstream/outlier_detection_impl.h:
```diff
@@ -99,50 +101,107 @@ class SuccessRateAccumulator {
 
+class SuccessRateMonitor {
+public:
+  SuccessRateMonitor() = delete;
```
nit: I don't think this is needed. It shouldn't compile if there is another constructor.

cpakulski replied:
Yes. Deleted that line.

mattklein123 commented on source/common/router/router.cc:
```diff
@@ -667,7 +671,8 @@ void Filter::onUpstreamReset(Http::StreamResetReason reset_reason,
                    Http::Utility::resetReasonToString(reset_reason));
 
-  updateOutlierDetection(Http::Code::ServiceUnavailable, upstream_request);
+  updateOutlierDetection(Upstream::Outlier::Result::CONNECT_FAILED, upstream_request,
```
In the non-split case don't you still need to track this as a 503? Or maybe that is done automatically?

cpakulski replied:
Done automatically in non-split mode:  https://github.com/cpakulski/envoy/blob/issue/3643/source/common/upstream/outlier_detection_impl.cc#L108-L109

mattklein123 commented on source/common/upstream/outlier_detection_impl.cc:
```diff
@@ -94,16 +119,96 @@ Http::Code DetectorHostMonitorImpl::resultToHttpCode(Result result) {
-  case Result::SERVER_FAILURE:
-    http_code = Http::Code::ServiceUnavailable;
+  case Result::CONNECT_SUCCESS:
+    ASSERT(false); // SUCCESS is used to report connection level success, not transaction
```
Use NOT_REACHED if this can't happen.

cpakulski replied:
Done.

mattklein123 commented on include/envoy/upstream/outlier_detection.h:
```diff
@@ -82,7 +90,7 @@ class DetectorHostMonitor {
    *         or the cluster did not have enough hosts to run through success rate outlier ejection.
    */
-  virtual double successRate() const PURE;
+  virtual double successRate(envoy::data::cluster::v2alpha::OutlierEjectionType) const PURE;
```
From an API perspective, the use of `OutlierEjectionType` here feels strange to me. Aren't you basically just splitting on external vs. internal and possibly returning a unified view if that is what is configured? Should we have a different enum that makes that more clear? Same elsewhere?

cpakulski replied:
Actually I used enum originally, but then refactored to use constants from API. Refactored back to enum.

mattklein123 commented on source/common/upstream/outlier_detection_impl.h:
```diff
@@ -99,50 +101,107 @@ class SuccessRateAccumulator {
+  // counters for local origin failures
+  std::atomic<uint32_t> consecutive_local_origin_failure_{0};
+
+  absl::flat_hash_map<envoy::data::cluster::v2alpha::OutlierEjectionType,
```
Do we need the hash map and related complexity? It seems like you can just have 2 stack allocated SR monitors and then a function that returns one of them depending on the config?

cpakulski replied:
Yes. Probably it was overkill. Allocated 2 instances of SR monitors and added simple method to return one of them depending on the type: https://github.com/cpakulski/envoy/blob/issue/3643/source/common/upstream/outlier_detection_impl.h#L158-L160

cpakulski commented:
/retest

mattklein123 commented on include/envoy/upstream/outlier_detection.h:
```diff
@@ -42,6 +42,9 @@ enum class Result {
 class DetectorHostMonitor {
 public:
+  // Types of Success Rate monitors.
+  using SuccessRateMonitorType = enum { externalOrigin, localOrigin };
```
nit: switch to `enum class`, also `ExternalOrigin` and `LocalOrigin`

cpakulski replied:
Changed to scoped _enum_. Thanks for catching it.

mattklein123 commented on source/common/upstream/outlier_detection_impl.cc:
```diff
@@ -37,14 +37,10 @@ DetectorHostMonitorImpl::DetectorHostMonitorImpl(std::shared_ptr<DetectorImpl> d
-      [envoy::data::cluster::v2alpha::OutlierEjectionType::SUCCESS_RATE_LOCAL_ORIGIN] =
-          std::make_unique<SuccessRateMonitor>(
-              envoy::data::cluster::v2alpha::OutlierEjectionType::SUCCESS_RATE_LOCAL_ORIGIN);
+  externalOriginSRMonitor_ = std::make_unique<SuccessRateMonitor>(
```
nit: snake case, same below, also this can be initialized in the initializer list above.

cpakulski replied:
Done. Moved initialization to the list.

cpakulski commented:
/retest

cpakulski commented:
/retest

mattklein123 commented on source/common/upstream/outlier_detection_impl.cc:
```diff
@@ -94,16 +110,94 @@ Http::Code DetectorHostMonitorImpl::resultToHttpCode(Result result) {
     break;
-  case Result::SERVER_FAILURE:
-    http_code = Http::Code::ServiceUnavailable;
+  case Result::CONNECT_SUCCESS:
```
It makes me a little uncomfortable that we are not handling an enum value in this way. Should the enums actually be split into local/external enums, and then how those are handled depends on whether splitting is on or not? WDYT?

cpakulski replied:
Processing of enums is actually split depending on the mode: (split vs non-split):
For non-split mode: https://github.com/envoyproxy/envoy/blob/4ba5f898e20f7f8e7a228c4ad78b81b43971a736/source/common/upstream/outlier_detection_impl.cc#L124-L125 and for split mode: https://github.com/envoyproxy/envoy/blob/4ba5f898e20f7f8e7a228c4ad78b81b43971a736/source/common/upstream/outlier_detection_impl.cc#L140-L141.

Also see my response to your comment titled: _Shouldn't this be CONNECT_SUCCESS?_

mattklein123 commented on source/common/upstream/outlier_detection_impl.cc:
```diff
@@ -94,16 +110,94 @@ Http::Code DetectorHostMonitorImpl::resultToHttpCode(Result result) {
+                                                            absl::optional<uint64_t> code) {
+  switch (result) {
+  case Result::CONNECT_SUCCESS:
+    // SUCCESS means that connection to host was established, but not transaction still can fail.
```
I'm a little confused by this. Why don't we track this result type in the non-split case? Maybe it's related to my other comment. Can you potentially rephrase this comment? Maybe that will help.

cpakulski replied:
See my response to your comment titled: _Shouldn't this be CONNECT_SUCCESS?_
I will also update the comment.

mattklein123 commented on source/common/upstream/outlier_detection_impl.cc:
```diff
@@ -94,16 +110,94 @@ Http::Code DetectorHostMonitorImpl::resultToHttpCode(Result result) {
+  // level failed. Since it it similar to HTTP 5xx, map it to 5xx handler.
+  case Result::REQUEST_FAILED:
+    // map it to http code and call http handler.
+    return putHttpResponseCode(500);
```
should we be using 503 here? Can we use `Code` enums here and below?

cpakulski replied:
Done: converted to enum class Http::Code.

cpakulski commented on source/common/tcp_proxy/tcp_proxy.cc:
```diff
@@ -535,7 +535,7 @@ void Filter::onUpstreamEvent(Network::ConnectionEvent event) {
 
     read_callbacks_->upstreamHost()->outlierDetector().putResult(
-        Upstream::Outlier::Result::SUCCESS);
+        Upstream::Outlier::Result::REQUEST_SUCCESS);
```
@mattklein123 No - it should be REQUEST_SUCCESS. I had a long debugging session when testing consecutive 5xx in non-split mode. For some reason they did not kick-in and the reason was using REQUEST_SUCCESS vs CONNECT_SUCCESS. I will try to explain it here. Feel free to ask for more clarification if this is still not clear. This will also address your other comments why CONNECT_SUCCESS is handled differently and not mapped to HTTP 200 OK in non-split mode.

For router I added two events: CONNECT_FAILED and CONNECT_SUCCESS. CONNECT_FAILED usage is clear: it is called when Envoy detects localOrigin error. 

For CONNECT_SUCCESS, the sequence of events is as follows. Let us consider the case when connection to upstream server is OK, but the server repeatedly returns 5xx code. Because connection is OK, router will report two events to OutlierDetector: CONNECT_SUCCESS and actual received HTTP code 5xx. If CONNECT_SUCCESS was mapped to HTTP 200 OK, then for each request the outlier detector would receive sequence of {200, 5xx}, which is obviously wrong. In spit-mode CONNECT_SUCCESS goes to consecutive_local_origin bucket and does not intervene with HTTP codes returned from upstream server, but in non-split mode it must not be mapped to 200 OK.

For cases when you want the event to be mapped to HTTP code 200, REQUEST_SUCCESS should be used. And this is the case here in tcp_proxy: CONNECT_FAILED is mapped to 5xx error and REQUEST_SUCCESS is mapped to 200 OK. Therefore, in non-split mode the code behaves as before and configuration is backwards compatible.

mattklein123 replied:
OK I see your point. I think what I'm saying is instead of having a single enum that tracks all errors, why not have 2 enums: `LocalOriginErrors` and `ExternalOriginErrors`. I think that will end up being a lot more clear through the code, regardless if split or not split is used?

cpakulski replied:
Let me think a bit about that. With 2 enum types there would be two interfaces to OutlierDetector. Let me go through the code and see how would it work.

cpakulski replied:
IMHO I do not think that having 2 APIs: one accepting LocalOriginErrors and second accepting ExternalOriginErrors would improve clarity much. Router reports HTTP codes through yet another API _putHTTPcode_ . The implementation of those APIs would be identical: map to HTTP code in non-split mode and call localOrigin methods for split mode. What do you think about renaming enum values we already have: instead of TIMEOUT, there would be LOCAL_ORIGIN_TIMEOUT, instead of REQUEST_SUCCESS there would be EXTERNAL_ORIGIN_REQUEST_SUCCESS, etc.

mattklein123 replied:
I think prefixing with LOCAL_ORIGIN and EXTERNAL_ORIGIN would help, but I still don't understand why this is REQUEST_SUCCESS. This is the tcp_proxy. Why are we tracking "requests" here? Shouldn't this be just connection success and failure that are all local origin?

/wait-any

cpakulski replied:
I agree with you that intuitively it should be CONNECT_SUCCESS. But it goes again to my response to your comment titled: _Shouldn't this be CONNECT_SUCCESS?_. The problem is that in non-split mode CONNECT_SUCCESS is not mapped to any HTTP code. (If it was mapped to 200 OK, then router would receive two HTTP codes for each request). But for tcp_proxy, in non-split mode, we have to make a call to outlier detector with a parameter which is mapped to 200 OK. And this is what REQUEST_SUCCESS is doing.
But wait, maybe I can add optional parameter when reporting events to Outlier Detector. TIMEOUT is already doing it. Maybe I can always use CONNECT_SUCCESS and add 200 OK in tcp_proxy and NIL in router. It would be much clearer.

mattklein123 replied:
+1. I think the other alternative here is to change the outlier detector APIs to make them work better for all the use cases we have now. Feel free to think about that. I think fundamentally this is problem is that it's too confusing to reason about what is going on with the names and APIs we have. Maybe approach it from that angle and see what falls out?

/wait-any

cpakulski replied:
I redesigned internal logic of reporting non-http results to outlier detector. In particular [putResult](https://github.com/cpakulski/envoy/blob/issue/3643/source/common/upstream/outlier_detection_impl.cc#L126-L139) method. LOCAL_ORIGIN_CONNECT_SUCCESS was previously handled in a specific way, which was hard to understand. Now, depending on the passed value of the _code_ there is no mapping to HTTP code, mapping based on defaults and mapping to passed HTTP code. The logic how to report events to outlier detector has been moved to individual filters, which in my opinion is much more clear.
As for API change to outlier detector, I am not sure whether much can be done without dropping some of the requirements, especially requirement to map non-http codes (LOCAL_ORIGIN_XXX and EXT_ORIGIN_XXX) to HTTP codes. Any sort of API would have to carry optional HTTP code, even when the code is later on not used in split mode.

cpakulski commented:
@mattklein123 I do not mind adding new changes until you feel comfortable with the code. If you feel that it brings too much complexity than it is worth I am OK with dropping this PR altogether. It could be implemented cleaner if we did not impose the requirement for backwards compatibility, but I understand why it is required.

mattklein123 commented:
> If you feel that it brings too much complexity than it is worth I am OK with dropping this PR altogether.

As long as you are willing to keep working on it, I think we should keep going. This is a useful feature for sure, I just want to make sure that we add it in a way that isn't super confusing. Thank you for your patience as we work through this. It's tough stuff and you are doing a great job!

cpakulski commented:
/retest

cpakulski commented on api/envoy/data/cluster/v2alpha/outlier_detection_event.proto:
```diff
@@ -52,7 +52,12 @@ enum OutlierEjectionType {
   // In case upstream host returns certain number of consecutive gateway errors
   CONSECUTIVE_GATEWAY_FAILURE = 1;
   // Runs over aggregated success rate statistics from every host in cluster
```
Updated with some explanations and linked to Architecture Overview::Outlier Detection page.

cpakulski commented:
> As long as you are willing to keep working on it, I think we should keep going.

No problem. Let us continue. It is great project.

cpakulski commented:
I addressed all comments. One thing I could still add, is some logic in source/server/http/admin.cc not to display local_origin_xxx values in non-split mode. The logic would be to check if the outlier detector is configured in spit mode and display _local_origin_consecutive_ and _local_origin_success_rate_ values. Otherwise those values would be omitted.

cpakulski commented:
/retest

cpakulski commented:
/retest

cpakulski commented:
/retest

cpakulski commented:
/retest

cpakulski commented:
@mattklein123 The code is pretty much finished. Changed some logic in outlier detector to make things more intuitive. Details and reasoning are here: https://github.com/envoyproxy/envoy/pull/4822#discussion_r286945505

mattklein123 commented:
Thanks @cpakulski will take a look soon.

mattklein123 commented:
@cpakulski sorry for the delay, I've been on vacation. Can you merge master and I will review again ASAP? Thank you.

/wait

mattklein123 commented on include/envoy/upstream/outlier_detection.h:
```diff
@@ -81,8 +93,14 @@ class DetectorHostMonitor {
    *         -1 means that the host did not have enough request volume to calculate success rate
    *         or the cluster did not have enough hosts to run through success rate outlier ejection.
+   * @param type specifies for which Success Rate Monitor the success rate value should be returned.
+   If the outlier detector is configured not to split external and local origin errors,
```
nit: missing '*' at beginning of block comment lines.

cpakulski replied:
Done.

mattklein123 commented on include/envoy/upstream/outlier_detection.h:
```diff
@@ -81,8 +93,14 @@ class DetectorHostMonitor {
+   * @param type specifies for which Success Rate Monitor the success rate value should be returned.
+   If the outlier detector is configured not to split external and local origin errors,
+   ExternalOrigin type returns success rate for all types of errors: external and local origin and
+   LocalOrigin type returns -1. If the outlier detector is configured to split external and local
```
Instead of returning -1 here, would it be better to use absl::optional<>? Maybe something to think about as a follow up.

cpakulski replied:
That -1 is displayed on GUI screen by admin.cc. If nobody interprets it, I can change it (would prefer another PR).

mattklein123 commented on source/common/router/router.cc:
```diff
@@ -771,7 +776,14 @@ void Filter::onUpstreamReset(Http::StreamResetReason reset_reason,
                    Http::Utility::resetReasonToString(reset_reason));
 
-  updateOutlierDetection(Http::Code::ServiceUnavailable, upstream_request);
+  /*
```
nit: use '//' comments.

cpakulski replied:
Done.

mattklein123 commented on source/common/router/router.cc:
```diff
@@ -1399,6 +1411,13 @@ void Filter::UpstreamRequest::onPoolReady(Http::StreamEncoder& request_encoder,
+  // If it was mapped to HTTP code it would interfere with code returned by HTTP server and two
+  // codes would be reported to outlier detector for one request.
+  host->outlierDetector().putResult(Upstream::Outlier::Result::LOCAL_ORIGIN_CONNECT_SUCCESS,
+                                    absl::optional<uint64_t>(0));
```
Would it be better to have a different putResult API that better captures this case? Sending 0 here seems like a hack to me. Why can't we just send nullopt, because that is the default? If so, maybe another API makes sense? It would be nice to match this up somehow w/ the LOCAL_ORIGIN_CONNECT_FAILED call above.

mattklein123 commented on source/common/upstream/outlier_detection_impl.cc:
```diff
@@ -33,6 +33,21 @@ DetectorSharedPtr DetectorImplFactory::createForCluster(
+                                                 HostSharedPtr host)
+    : detector_(detector), host_(host),
+      // add Success Rate monitors
+      external_origin_SR_monitor_(std::make_unique<SuccessRateMonitor>(
```
Can these just be member variables? Do they need to actually be pointers and allocated additionally?

cpakulski replied:
Done. Changed them to be member variables.

mattklein123 commented on source/common/upstream/outlier_detection_impl.cc:
```diff
@@ -82,28 +98,111 @@ Http::Code DetectorHostMonitorImpl::resultToHttpCode(Result result) {
+// are not treated differently. All errors are mapped to HTTP codes.
+// Depending on the value of the parameter *code* the function behaves differently:
+// - if the *code* is not defined, mapping uses resultToHttpCode method to do mapping.
+// - if *code* is zero, no mapping takes place: result is dropped and not reported to outlier
```
Per my comment I would like to figure out a better API to avoid this 0 hack. Thoughts?

cpakulski replied:
Currently putResult is used in 3 scenarios:
- report event with optional HTTP code. The specified HTTP code is directly fed to outlier detector without any conversions.
- report event without optional HTTP code. OutlierDetector uses legacy code https://github.com/cpakulski/envoy/blob/issue/3643/source/common/upstream/outlier_detection_impl.cc#L97-L117 to convert the event to HTTP code based on defaults, which is passed to outlier detector.
- report event, but do not convert it to any HTTP code. This is to catch scenario with router, which must not convert LOCAL_ORIGIN_CONNECT_SUCCESS to HTTP code 200. Otherwise it will collide with the response from the real HTTP server.

I somehow have to indicate not to convert to HTTP code to handle scenario 3. nullopt is already used for scenario 2, what means " I am not giving any code, use defaults for conversion".
I think that the cleanest way would be to always require to specify HTTP code in API. So instead of writing:
`putResult(EXT_ORIGIN_REQUEST_SUCCESS)`
one would have to write
`putResult(EXT_ORIGIN_REQUEST_SUCCESS, absl::optional<uint64_t>(200))`
Then specifying nullopt would mean "Do nothing with HTTP code" to handle scenario 3.

I would have to convert all instances of putResult to add the code - about 8-10 places.

WDYT?

mattklein123 replied:
Fundamentally I think the issue we have is this API is broken and we are reaching the limits of what we can do with it. Why do we need HTTP codes for non-HTTP paths? Can we have a the following APIs:

`putHttpResult(...)` (for this one just make the caller provide the code each time)
`putExternalOriginResult(...)` (maybe this just takes success/failure like for redis)
`putLocalOriginResult(...)`

And then can't we just do the right thing with them?

cpakulski replied:
>  Why do we need HTTP codes for non-HTTP paths?

We need them because of backward compatibility. They were previously mapped to HTTP codes.

> putHttpResult(...) (for this one just make the caller provide the code each time)
> putExternalOriginResult(...) (maybe this just takes success/failure like for redis)
> putLocalOriginResult(...)

There is no problem with implementing/using putHttpResult and putExternalOriginResult. They always eventually go to external origin counters (consecutive_5xx) regardless whether we work in split or non-split mode.

The problem is with putLocalOriginResult. I went through all the cases where LOCAL_ORIGIN event is reported:
- for LOCAL_ORIGIN_TIMEOUT, the code actually instructs to which HTTP code it should be translated: https://github.com/cpakulski/envoy/blob/issue/3643/source/common/router/router.cc#L612-L616, so HTTP code must be passed to outlier detector.
- for LOCAL_ORIGIN_CONNECT_SUCCESS: tcp_proxy wants to translate it to code 200 so it behaves as before my changes:https://github.com/cpakulski/envoy/blob/issue/3643/source/common/tcp_proxy/tcp_proxy.cc#L539-L540. However, router does not want to translate CONNECT_SUCCESS to HTTP code 200 because it will report the real code via putHttpResult. So here to indicate that no conversion should happen, I stick value of zero: https://github.com/cpakulski/envoy/blob/issue/3643/source/common/router/router.cc#L1414-L1419

I can implement 

> putHttpResult(...) (for this one just make the caller provide the code each time)
> putExternalOriginResult(...) (maybe this just takes success/failure like for redis)
> putLocalOriginResult(...)

but putLocalOriginResult must carry HTTP code and somehow I have to indicate not to convert to any code to handle router's case. One way would be to modify include/envoy/http/codes.h and add enum InvalidCode = 0, so we do not hack it explicitly with zero, but use valid enum.

If this sounds good I will implement this today.

mattklein123 replied:
Sorry this is making my head hurt. Can we somehow unify the TCP proxy and router local origin cases so we don't need the special case, even if it is not explicitly the exact same behavior?

cpakulski replied:
I understand that this is mind boggling. It is hard to design nice and elegant API to serve all the cases. There is no way to simply unify TCP cases and router cases. I somehow have to distinguish them. In order to avoid a hack with passing value zero I added another event LOCAL_ORIGIN_CONNECT_SUCCESS_FINAL to indicate that there is no other protocol which will report final error/success.

cpakulski replied:
I also experimented with implementing 
```
putHttpResult(...) (for this one just make the caller provide the code each time)
putExternalOriginResult(...) (maybe this just takes success/failure like for redis)
putLocalOriginResult(...)
```
but outlier detector's API is most often invoked through wrappers, so I  would have to implement a wrapped for each of those methods. It very quickly became ugly.

mattklein123 commented on source/common/upstream/outlier_detection_impl.cc:
```diff
@@ -82,28 +98,111 @@ Http::Code DetectorHostMonitorImpl::resultToHttpCode(Result result) {
+  put_result_func_(this, result, code);
+}
+
+void DetectorHostMonitorImpl::localOriginFailure() {
```
Is it possible to share this code somehow with the other failure function? There is some tricky logic here and I'm wondering if somehow there could be a shared function that is used in both places? Not sure if that is possible or not but just throwing it out there for thought.

cpakulski replied:
Not sure what you mean. I went through the code and does not see how they could be shared.

cpakulski commented:
/retest

cpakulski commented:
@mattklein123 rebase is completed. Thanks for reviewing my latest changes.

cpakulski commented:
@mattklein123 rebase is completed. Thanks for reviewing my latest changes.

cpakulski commented:
Thanks @mattklein123 for your patience and valuable comments.

I added many lines of documentaion regarding outlier detector. Most likely issue https://github.com/envoyproxy/envoy/issues/3730 can also be closed.
